### PR TITLE
Added support for MPEG-SD interactivity extensions

### PR DESCRIPTION
**Pull request by Etiennefaivredarcier**
This pull request is related to an implementation of the MPEG-SD interactivity extensions.

It requires the import of the package: com.unity.inputsystem (from Unity, via the package manager: Window->Package Manager). It has been tested on the version 1.4.2.

This implementation supports the parsing of a glTF file with the whole interactivity model with the following triggers:

- TRIGGER_COLLISION
- TRIGGER_PROXIMITY
- TRIGGER_USER_INPUT
- TRIGGER_VISIBILITY

And the following actions:

- ACTION_ACTIVATE
- ACTION_TRANSFORM
- ACTION_BLOCK
- ACTION_ANIMATION
- ACTION_MEDIA
- ACTION_MANIPULATE
- ACTION_SET_MATERIAL
- ACTION_HAPTIC
- ACTION_AVATAR

There are a few details of the implementations that are not (yet) released:

- Primitives are supported, but for collision they are using a mesh collider model, they should use a more performant way to process
- The ACTION_USER_INPUT support keyboard and mouse for now, it should also being capable to handle real XR devices
- Pause functionality in the ACTION_ANIMATION is not implemented
- Behavior priority is not implemented
- ACTION_HAPTIC is not implemented
- ACTION_MEDIA is not implemented
- ACTION_AVATAR is not implemented

### CODE REVIEW

**Comment by nlsdvl** [comment]
@Etiennefaivredarcier, Thank you for this contribution. I'm progressing the code review, I'll add comments to the [rt-xr-content](https://github.com/5G-MAG/rt-xr-content/pull/4) repo as well.

In the PR description here, you mention a dependency on _com.unity.inputsystem 1.4.2_.

Can you please specify which version of the dependency is needed by adding it to the [package.json](https://docs.unity3d.com/Manual/upm-manifestPkg.html) file dependencies declaration instead ?

I tried _com.unity.inputsystem 1.4.2_ without luck. 
In my experience however version **1.7.0** was working out of the box. 

I found the code compiled depending on definition of `USE_NEW_INPUT_SYSTEM`. Based on the dependency declaration in package.json, is this define still relevant ?

**Comment by Etiennefaivredarcier** [comment]
> @Etiennefaivredarcier, Thank you for this contribution. I'm progressing the code review, I'll add comments to the [rt-xr-content](https://github.com/5G-MAG/rt-xr-content/pull/4) repo as well.
> 
> In the PR description here, you mention a dependency on _com.unity.inputsystem 1.4.2_.
> 
> Can you please specify which version of the dependency is needed by adding it to the [package.json](https://docs.unity3d.com/Manual/upm-manifestPkg.html) file dependencies declaration instead ?
> 
> I tried _com.unity.inputsystem 1.4.2_ without luck. In my experience however version **1.7.0** was working out of the box.
> 
> I found the code compiled depending on definition of `USE_NEW_INPUT_SYSTEM`. Based on the dependency declaration in package.json, is this define still relevant ?

Hello @nlsdvl ,

Thank you for reviewing,

I just removed my Input system package from the package manager and inserted it by hand in the package.json, and it works on my side. Could you please make sure that you added it inside the dependencies object like the following ?

`  "dependencies": {`
`    "com.unity.modules.jsonserialize": "1.0.0",`
`    "com.unity.modules.unitywebrequest": "1.0.0",`
`    "com.unity.inputsystem": "1.4.2",`
`    "com.unity.mathematics": "1.2.6",`
`    "com.unity.burst": "1.6.6"`
`  },`

Could you also make sure that you put USE_NEW_INPUT_SYSTEM inside the Scripting Define Symbols ?

Also, the code compilation depends on this package because (for now) because I prefer to not make assumption on the package a developer want to use. Of course, as there is only one implementation, it doesn't make sense (yet). Do you prefer that we get rid of the compilation definition?

File: Runtime/Scripts/glTFast.asmdef
Comment (nlsdvl) [inline Runtime/Scripts/glTFast.asmdef]: @Etiennefaivredarcier 
'USE_NEW_INPUT_SYSTEM' evaluates to true when 'com.unity.inputsystem' >= 1.5 is installed, but evaluates to false when v1.4.2 is installed.

File: Runtime/Scripts/ActionManipulate.cs
Comment (nlsdvl) [inline Runtime/Scripts/ActionManipulate.cs:61]: InputAction class is not defined when USE_NEW_INPUT_SYSTEM evaluates to false at compile time. the program doesn't compile when USE_NEW_INPUT_SYSTEM is false.

File: Runtime/Scripts/ActionManipulate.cs
Comment (nlsdvl) [inline Runtime/Scripts/ActionManipulate.cs:67]: InputAction class is not defined when USE_NEW_INPUT_SYSTEM evaluates to false at compile time. the program doesn't compile when USE_NEW_INPUT_SYSTEM is false.

**Comment by nlsdvl** [comment]
@Etiennefaivredarcier, indeed I saw USE_NEW_INPUT_SYSTEM is not fully implemented, and I don't know when that would happen, so I would prefer removing it and opening a ticket about it as a feature request.

> Could you also make sure that you put USE_NEW_INPUT_SYSTEM inside the Scripting Define Symbols ?

It is already defined in your pull request. The package's 'glTFast.asmdef' file has 'Version Defines' with 'USE_NEW_INPUT_SYSTEM' >= 1.5. It means 'USE_NEW_INPUT_SYSTEM' evaluates to true when 'com.unity.inputsystem' >= 1.5 is installed, but evaluates to false when v1.4.2 is installed. 

It causes theses compilation errors on my side when I use 1.4.2, as the type 'InputAction' is only imported when 'USE_NEW_INPUT_SYSTEM' evaluates to true:

> Packages\com.atteneder.gltfast\Runtime\Scripts\ActionManipulate.cs(65,38): error CS0246: The type or namespace name 'InputAction' could not be found (are you missing a using directive or an assembly reference?)
> Packages\com.atteneder.gltfast\Runtime\Scripts\ActionManipulate.cs(71,39): error CS0246: The type or namespace name 'InputAction' could not be found (are you missing a using directive or an assembly reference?)

I understand that you want to avoid hard restrictions on the dependency version. 
Unfortunately Unity's package.json dependency declaration doesn't help with that.
As far as I know it doesn't support specifying a version range (eg. >=1.5).

When I use version 1.5 or above it compiles as expected. 
I would recommend 1.7 as it is the most recent stable release of that package.

File: Runtime/Scripts/ActionTransform.cs
Comment (nlsdvl) [inline Runtime/Scripts/ActionTransform.cs:25]: this throws exception when ACTION_TRANSFORM is parsed.

Exception should be replaced with something like:

```
        public float Delay => m_Delay;
        private float m_Delay;
```

File: Runtime/Scripts/Schema/MpegSceneInteractivity.cs
Comment (nlsdvl) [inline Runtime/Scripts/Schema/MpegSceneInteractivity.cs]: Matrix4x4 is not parsed. Instead, an array of float should be parsed and manually converted to Matrix4x4: , see for instance https://github.com/atteneder/glTFast/blob/main/Runtime/Scripts/Schema/Node.cs#L50

File: Runtime/Scripts/Schema/MpegNodeInteractivity.cs
Comment (nlsdvl) [inline Runtime/Scripts/Schema/MpegNodeInteractivity.cs]: [Serializable] decorates the trigger class.
The node interactivity is not parsed. 
I tried using gravity to trigger collision without luck. The collision definition on the node is ignored.

**Comment by Etiennefaivredarcier** [comment]
Hello @nlsdvl ,
Thank you for your message - I've commited a more restrictive approach by removing the define USE_NEW_INPUT_SYSTEM since, it's probably gonna be used for an XR experience anyway.

**Comment by nlsdvl** [comment]
Hello @Etiennefaivredarcier ,

Thank you for fixing the USE_NEW_INPUT_SYSTEM define. 

There are some issues pending in the pull request review, I suggest we merge and open tickets if you can not resolve them in the short term, especially:
- an issue with parsing MPEG_node_interactivity which breaks its implementation. It seems the physics implementation would otherwise work.
- an issue with parsing ACTION_TRANSFORM 4x4 matrix transform.
- an issue with ACTION_BLOCK which never stops after initial trigger.

While reviewing, I also produced some additional test content to be able to QA the PR, please see my comments on the content repo.

**Comment by Etiennefaivredarcier** [comment]
Hello @nlsdvl,

Thank you for your message,
I agree and suggest that you merge this PR if you can.
I've created the issues in the Github project, don't hesitate to add further informations or issues you encouter.
Could you also elaborate on the physics implementation issue you had ?

Thank you

**Comment by nlsdvl** [comment]
@Etiennefaivredarcier as discussed offline, we would prefer to have these resolved before merging.

**Comment by nlsdvl** [comment]
Hi @Etiennefaivredarcier thanks for updating the PR. 
The code looks good. I'll be processing it shortly.

**Review by nlsdvl — APPROVED** [review]
Thank you for addressing my comments. 
I am also processing the [PR#4](https://github.com/5G-MAG/rt-xr-content/pull/4) for the test content.

**Pull request merged**
